examples/bloom-embeddings.ipynb

```diff
@@ -0,0 +1,729 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This gives the model a little more resolution for the unknown words. If all\n",
+    "out-of-vocabulary words are assigned the same vector, then they'll all look\n",
+    "identical to the model. Even if the training data actually includes information\n",
+    "that shows two different out-of-vocabulary words have important, different\n",
+    "implications -- for instance, if one word is a strong indicator of positive\n",
+    "sentiment, while the other is a strong indicator of negative sentiment -- the\n",
+    "model won't be able to tell them apart. However, if we have 10 buckets for the\n",
+    "unknown words, we might get lucky, and assign these words to different buckets.\n",
+    "If so, the model would be able to learn that one of the unknown-word vectors\n",
+    "makes positive sentiment more likely, while the other vector makes negative\n",
+    "sentiment more likely.\n",
+    "\n",
+    "If this is good, then why not do more of it? Bloom embeddings are like an\n",
+    "extreme version, where _every_ word is handled like the unknown words above:\n",
+    "there are 100 vectors for the \"unknown\" portion, and 0 for the \"known\" portion.\n",
+    "\n",
+    "So far, this approach seems weird, but not necessarily good. The part that makes\n",
+    "it unfairly effective is the next step: by simply doing the same thing multiple\n",
+    "times, we can greatly improve the resolution, and have unique representations\n",
+    "for far more words than we have vectors. The code in full:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "executionInfo": {
+     "elapsed": 8,
+     "status": "ok",
+     "timestamp": 1648197799885,
+     "user": {
+      "displayName": "Vincent D. Warmerdam",
+      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14Gh4KYzhhhK0YDTnAQsUIaQPw-0dKIP-kLBID7nFdQ=s64",
+      "userId": "05641618555626735638"
+     },
+     "user_tz": -60
+    },
+    "id": "tTkqM3EixhWM"
+   },
+   "outputs": [],
+   "source": [
+    "import numpy\n",
+    "import mmh3\n",
+    "\n",
+    "def allocate(n_vectors, n_dimensions):\n",
+    "    table = numpy.zeros((n_vectors, n_dimensions), dtype='f')\n",
+    "    table += numpy.random.uniform(-0.1, 0.1, table.size).reshape(table.shape)\n",
+    "    return table\n",
+    "\n",
+    "def get_vector(table, word):\n",
+    "    hash1 = mmh3.hash(word, seed=0)\n",
+    "    hash2 = mmh3.hash(word, seed=1)\n",
+    "    row1 = hash1 % table.shape[0]\n",
+    "    row2 = hash2 % table.shape[0]\n",
+    "    return table[row1] + table[row2]\n",
+    "\n",
+    "def update_vector(table, word, d_vector):\n",
+    "    hash1 = mmh3.hash(word, seed=0)\n",
+    "    hash2 = mmh3.hash(word, seed=1)\n",
+    "    row1 = hash1 % table.shape[0]\n",
+    "    row2 = hash2 % table.shape[0]\n",
+    "    table[row1] -= 0.001 * d_vector\n",
```

Comment:
Is this `0.001` the learning-rate? Would it make sense to have a `learn_rate` parameter here?

Reply:
I don't think this is needed for this example.